pkg/planner/core/casetest/vectorsearch/testdata/ann_index_suite_out.json

```diff
@@ -507,5 +507,279 @@
+        ],
+        "Warn": null
+      },
+      {
+        "SQL": "explain select * from t1 order by vec_cosine_distance(vec, '[1,1,1]') limit 10",
+        "Plan": [
+          "Projection_6 10.00 root  test.t1.id, test.t1.vec, test.t1.a, test.t1.b, test.t1.c, test.t1.d",
+          "└─TopN_10 10.00 root  Column#8, offset:0, count:10",
+          "  └─TableReader_27 10.00 root  MppVersion: 2, data:ExchangeSender_26",
+          "    └─ExchangeSender_26 10.00 mpp[tiflash]  ExchangeType: PassThrough",
+          "      └─TopN_25 10.00 mpp[tiflash]  Column#8, offset:0, count:10",
+          "        └─Projection_24 10.00 mpp[tiflash]  test.t1.id, test.t1.vec, test.t1.a, test.t1.b, test.t1.c, test.t1.d, vec_cosine_distance(test.t1.vec, [1,1,1])->Column#8",
+          "          └─TableFullScan_23 10.00 mpp[tiflash] table:t1, index:idx_embedding(vec) keep order:false, annIndex:COSINE(vec..[1,1,1], limit:10)"
+        ],
+        "Warn": null
+      },
+      {
+        "SQL": "explain select id, a, b from t1 order by vec_cosine_distance(vec, '[1,1,1]') limit 10",
+        "Plan": [
+          "Projection_7 10.00 root  test.t1.id, test.t1.a, test.t1.b",
+          "└─TopN_11 10.00 root  Column#11, offset:0, count:10",
+          "  └─TableReader_28 10.00 root  MppVersion: 2, data:ExchangeSender_27",
+          "    └─ExchangeSender_27 10.00 mpp[tiflash]  ExchangeType: PassThrough",
+          "      └─TopN_26 10.00 mpp[tiflash]  Column#11, offset:0, count:10",
+          "        └─Projection_25 10.00 mpp[tiflash]  test.t1.id, test.t1.vec, test.t1.a, test.t1.b, vec_cosine_distance(test.t1.vec, [1,1,1])->Column#11",
+          "          └─TableFullScan_24 10.00 mpp[tiflash] table:t1, index:idx_embedding(vec) keep order:false, annIndex:COSINE(vec..[1,1,1], limit:10)"
+        ],
+        "Warn": null
+      },
+      {
+        "SQL": "explain select a, id, b from t1 order by vec_cosine_distance(vec, '[1,1,1]') limit 10",
+        "Plan": [
+          "Projection_7 10.00 root  test.t1.a, test.t1.id, test.t1.b",
+          "└─TopN_11 10.00 root  Column#11, offset:0, count:10",
+          "  └─TableReader_28 10.00 root  MppVersion: 2, data:ExchangeSender_27",
+          "    └─ExchangeSender_27 10.00 mpp[tiflash]  ExchangeType: PassThrough",
+          "      └─TopN_26 10.00 mpp[tiflash]  Column#11, offset:0, count:10",
+          "        └─Projection_25 10.00 mpp[tiflash]  test.t1.id, test.t1.vec, test.t1.a, test.t1.b, vec_cosine_distance(test.t1.vec, [1,1,1])->Column#11",
+          "          └─TableFullScan_24 10.00 mpp[tiflash] table:t1, index:idx_embedding(vec) keep order:false, annIndex:COSINE(vec..[1,1,1], limit:10)"
+        ],
+        "Warn": null
+      },
+      {
+        "SQL": "explain select id, vec_cosine_distance(vec, '[1,1,1]') as d from t1 order by d limit 10",
+        "Plan": [
+          "Projection_6 10.00 root  test.t1.id, vec_cosine_distance(test.t1.vec, [1,1,1])->Column#7",
```

Comment:
Remember to open an issue in TiDB and mark this as an optimization to do.

Reply:
Issue #58217 has been opened.

Reply:
This is not what #56318 is talking about. This is another redundent projection when the function is directly required in the SELECT clause.

Reply:
issue #58242 has opened.